### DISCUSSION TITLE
Thonny's debug feature "issue"

### DISCUSSION THREAD

**Discussion by villares** [tabreturn/thonny-py5mode#61] [Ideas]
This looks like a really hard thing: 
Thonny's debug feature doesn't work with "imported mode" and I was reminded of that by a recent (unrelated) -Processing forum post. I wonder if there is a solution to this or it is just something we might want to warn users about when they try "the bug button".

I think Aivar had commented  in a conversation we had, a long time ago, on this being a huge drawback of having a "simplified language mode pre-processing"...

**Comment by hx2A** [comment]
Right. Since imported mode uses the run_sketch command line tool, any debugging would have to be done through that. I did learn some things about debuggers while at PyCon. It would be very difficult to add debugging or the ability to step through code as a feature in run_sketch.

In module mode, debugging should work just fine but last time I tried that in VSCode it did not work for me. It was a while ago though. Does it work in Thonny?

It might be an interesting educational tool to debug a Sketch using the default renderer and see the Sketch being drawn one line at a time. Does that actually work? With the OpenGL renderers, the code could perhaps work with a debugger but you would not see the Sketch window change until after the `draw()` function was complete.

**Reply by villares** [reply]
The Thonny debug in "module mode", was way less useful than what I remembered :( 
You see it go through global scope, function definitions and when `py5.run_sketch()` is called nothing is seen afterwards.

**Reply by hx2A** [reply]
> The Thonny debug in "module mode", was way less useful than what I remembered :( You see it go through global scope, function definitions and when `py5.run_sketch()` is called nothing is seen afterwards.

For non-py5 Python code it probably works a lot better. Recall that py5 works by making calls _from_ Java _to_ Python. Every execution of `draw()` is called by Java. So what happens when you run a Sketch is that py5 collects the relevant Sketch functions, sets things up with a Processing Java Sketch, and then launches the Java Sketch. As far as the debugger is concerned, it steps through the code until the call to Processing's `runSketch()` method but does not see that return. On the Java side it is launching threads and one of those threads makes calls back to Python.

I do believe it would be possible for a debugger to work with py5. I know in VSCode the debugger can work with a Python website which also has a different way of executing Python code. I haven't gotten around to looking at that though.

**Reply by villares** [reply]
Today I tried a Pdb `breakpoint()` with module mode in Thonny, just out of curiosity. It might be useful...

![image](https://github.com/user-attachments/assets/019287d6-fbdc-4b41-9ad9-3b064ad26915)

(works on imported mode too)

**Reply by hx2A** [reply]
Neat! It never occurred to me to use `breakpoint()` in a running Sketch. I just tried it using what I had in front of me, which happens to be a new live coding feature...I didn't think it would work, but it actually worked perfectly. I didn't think the two would cooperate, but it worked great.
